rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+    dataset, num_consumers=1, num_tasks=None, deterministic=False):
+  """Begins distributed iteration over a dataset.
+
+  It is expected that the dataset contains at least one `.distribute(address)`
+  transformation, otherwise this method will print a warning and do nothing.
+
+  `create_iteration` will first register the dataset with the tf.data service
+  if it isn't already registered. It will then request the creation of
+  `num_consumers` dataset iterators which divide the dataset `num_consumers`
+  ways. The returned object can be used to read from one of the
+  iterators using
+  `tf.data.experimental.service.make_iterator(ds, obj, consumer_index)`.
+
+  ds = ... # dataset to distribute
+  ds = ds.apply(tf.data.experimental.service.distribute(address))
+  if consumer_index == 0:
+    # The iteration object is a byte array which needs to be shared among all
+    # consumers. Here we suppose there are broadcast_send and broadcast_recv
+    # method available.
+    iteration_id = tf.data.experimental.service.create_iteration(ds, address, 3)
+    broadcast_send(iteration_id)
+  else:
+    iteration_id = broadcast_recv()
+  it = tf.data.experimental.service.make_iterator(
+      ds, iteration_id, consumer_index)
+  for element in it:
+    # process element
+
+  Args:
+    dataset: The dataset to begin iteration over.
+    num_consumers: The number of consumers to divide the dataset between. Set
+      this if you require determinism. If None, a single iterator id is returned,
+      and any number of consumers can read from that iterator id. The data
+      produced by the dataset will be fed to consumers on a first-come
+      first-served basis.
+    num_tasks: The number of tasks to use for processing. Tasks run for
+      the duration of an epoch, and each worker should typically process a single
+      task. Normally it is best to leave this as None so that the master can
+      choose a reasonable number of tasks. Setting `num_tasks` is useful for
+      producing deterministic results.
+    deterministic: Whether the iteration should be performed
+      deterministically. Fully deterministic output also requires setting
+      `num_tasks` to a fixed number, and that the input dataset is itself
+      deterministic.
+
+  Returns:
+    An iteration_id which can be used to created iterators via
+      `tf.data.experimental.service.make_iterator`
+  """
+
+def tf.data.experimental.service.make_iterator(
+    dataset, iteration, consumer_index):
+  """Creates an iterator for reading from the specified dataset.
+
+  Args:
+    dataset: The dataset to read from.
+    iteration: An iteration_id object generated by
+      `tf.data.experimental.service.create_iteration`.
+    consumer_index: The consumer index within the iteration to read from. If
+      the iteration was created with `n` consumers, `consumers_index` must be
+      less than `n`.
+
+  Returns:
+    A Python iterator which iterates over the dataset elements.
+  """
+```
+
```

Comment:
An earlier discussion about this involved a public primitive to create a dataset from an iterator, which would also be useful internally when implementing create_iteration. Is that still in scope?

I ask because while mostly unrelated to the serve issue this would be very useful when dealing with the types of pipelines which are non-goals of the service project (stateful pipelines, pipelines with py_func, etc)

Reply:
Such an iterator-to-dataset operation would be useful for the implementation of `make_iterator`, but for the purpose of this design we would like to avoid publicly exposing any more API than we absolutely need to.

Reply:
Previously we were thinking that `distribute` would return an opaque object which can be used to create an iterator (or multiple shared iterators). Then to apply local transformations, we would need a way to go from iterator to dataset.

We decided to take an alternative path: the `distribute` transformation will return a regular `Dataset`, which the user can chain transformations onto. We no longer need to go from iterator to dataset.